src/common/oclengine.cpp

```diff
@@ -125,81 +235,44 @@ void OCLEngine::InitOCL()
         std::shared_ptr<OCLDeviceContext> devCntxt = std::make_shared<OCLDeviceContext>(
             devPlatVec[i], all_devices[i], all_contexts[all_contexts.size() - 1], plat_id);
 
-        cl::Program program = cl::Program(devCntxt->context, sources);
+        std::string fileName = "qrack_ocl_dev_" + std::to_string(i) + ".ir";
+        std::string clBinName = home + fileName;
+
+        std::cout << "Device #" << i << ", ";
+        cl::Program program = MakeProgram(buildFromSource, sources, clBinName, devCntxt);
 
         cl_int buildError = program.build({ all_devices[i] }, "-cl-denorms-are-zero -cl-fast-relaxed-math");
         if (buildError != CL_SUCCESS) {
             std::cout << "Error building for device #" << i << ": " << buildError << ", "
                       << program.getBuildInfo<CL_PROGRAM_BUILD_STATUS>(all_devices[i])
                       << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(all_devices[i]) << std::endl;
 
-            // The default device was set above to be the last device in the list. If we can't compile for it, we use
-            // the first device. If the default is the first device, and we can't compile for it, then we don't have any
-            // devices that can compile at all, and the environment needs to be fixed by the user.
+            // The default device was set above to be the last device in the list. If we can't compile for it, we
+            // use the first device. If the default is the first device, and we can't compile for it, then we don't
+            // have any devices that can compile at all, and the environment needs to be fixed by the user.
             if (i == dev) {
```

Comment:
Oh, this is an important question.  Is it feasible that you might have to switch between devices for compiled versions?  Is that supported via the InitOCL call in some fashion?

Reply:
Here's the user story:

1. Optionally, set the default precompilation saving/loading path at compile time.
2. Run `qrack_cl_precompile`, (optionally specifying a _different_ location by command line parameter,) to compile and save all OCL programs.
3. If you want some devices on JIT and some on precompiled programs, delete (or move) the binaries for the devices you want to use JIT. (I see no real benefit in this, except to clean out corrupt binaries, or to turn off default precompilation entirely.)
4. At run time, you can turn JIT on for all devices, or off for the full set of devices with precompiled binaries. You can also specify any path to attempt loading from.

For the complexity this adds to the initialization method, I see no benefit in it. This is still highly configurable, to meet that particular need, just not at run time.

Reply:
Ah so this process stores compiled binaries for all of the systems that are currently present/enabled?  Is there any chance where the binaries could be used for the wrong class of device?

Reply:
The only chance for loading the wrong binary is a hardware configuration change, such as installing a new device or switching card slots. In that case, I would always advise the user to run `qrack_cl_precompile` again, which should be sufficient to adjust for the change. (It's better to delete the cache files, first, but not strictly necessary.)

Reply:
In a virtual OpenCL environment, yes, this could be a hassle. However, we also fall back to JIT if the "build" after binary loading fails, which it would if the device was incompatible.

We give the user plenty of freedom and information to control this properly, but we can't do their hardware configuration for them.